File: Baekjoon OJ/16234 - 인구 이동.py
"""
    Code written on Feb 10, 2021
    출처: Baekjoon OJ, https://www.acmicpc.net
    16234번: 인구 이동
    link: https://www.acmicpc.net/problem/16234
"""
from collections import deque
from sys import stdin


def bfs(i, j, population):
    global borderline_open
    queue = deque([(i, j, population)])
    union = [(i, j)]
    total_population = population
    num_countries = 1
    while queue:
        y, x, population = queue.popleft()
        for k in range(4):
            ny = y + dy[k]
            nx = x + dx[k]
            if 0 <= nx < n and 0 <= ny < n and not visited[ny][nx] \
                    and l <= abs(population - countries[ny][nx]) <= r:
                queue.append((ny, nx, countries[ny][nx]))
                union.append((ny, nx))
                total_population += countries[ny][nx]
                num_countries += 1
                visited[ny][nx] = True
    if num_countries != 1:
        new_valid_locations.extend(union)
        borderline_open = True
        population = total_population // num_countries
        for y, x in union:
            countries[y][x] = population


n, l, r = map(int, stdin.readline().split())
countries = [list(map(int, stdin.readline().split())) for _ in range(n)]

valid_locations = ((i, j) for i in range(n) for j in range(n))
dy = [0, 0, 1, -1]
dx = [1, -1, 0, 0]
movement = 0
while True:
    visited = [[False] * n for _ in range(n)]
    new_valid_locations = []
    borderline_open = False
    for i, j in valid_locations:
        if not visited[i][j]:
            visited[i][j] = True
            bfs(i, j, countries[i][j])
    valid_locations = new_valid_locations
    if not borderline_open:
        break
    movement += 1
print(movement)

File: Baekjoon OJ/16234 - 인구 이동.java
/*
    Code written on May 28, 2021
    출처: Baekjoon OJ, https://www.acmicpc.net
    16234번: 인구 이동
    link: https://www.acmicpc.net/problem/16234
*/
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.StringTokenizer;

public class Main {
    private static int[][] map;
    private static int width;
    private static int minPopulation;
    private static int maxPopulation;

    public static void main(String[] args) throws Exception {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer tokenizer = new StringTokenizer(reader.readLine());
        width = Integer.parseInt(tokenizer.nextToken());
        minPopulation = Integer.parseInt(tokenizer.nextToken());
        maxPopulation = Integer.parseInt(tokenizer.nextToken());
        map = new int [width][width];

        for (int row = 0; row < width; row++) {
            tokenizer = new StringTokenizer(reader.readLine());
            for (int col = 0; col < width; col++)
                map[row][col] = Integer.parseInt(tokenizer.nextToken());
        }
        System.out.print(countNumOfMoves());
    }

    public static int countNumOfMoves() {
        int totalNumOfMoves = 0;
        while (true) {
            boolean[][] visited = new boolean[width][width];
            int numOfMoves = 0;
            for (int row = 0; row < width; row++) {
                for (int col = 0; col < width; col++) {
                    if (visited[row][col])
                        continue;
                    if (isBorderOpen(new Coordinate(row, col), visited))
                        numOfMoves = 1;
                }
            }
            if (numOfMoves == 0)
                return totalNumOfMoves;
            totalNumOfMoves += numOfMoves;
        }
    }

    public static boolean isBorderOpen(Coordinate nation, boolean[][] visited) {
        ArrayDeque<Coordinate> queue = new ArrayDeque<>();
        ArrayDeque<Coordinate> union = new ArrayDeque<>();
        int totalPopulation = 0;
        int numOfNations = 0;
        boolean isOpen = false;

        queue.add(nation);
        while (!queue.isEmpty()) {
            Coordinate coordinate = queue.pollFirst();
            int row = coordinate.getRow();
            int col = coordinate.getCol();
            if (visited[row][col])
                continue;
            visited[row][col] = true;
            union.add(coordinate);
            totalPopulation += map[row][col];
            numOfNations += 1;
            if (0 < row) {
                int populationDiff = Math.abs(map[row - 1][col] - map[row][col]);
                if (minPopulation <= populationDiff && populationDiff <= maxPopulation)
                    queue.addLast(new Coordinate(row - 1, col));
            }
            if (row < width - 1) {
                int populationDiff = Math.abs(map[row + 1][col] - map[row][col]);
                if (minPopulation <= populationDiff && populationDiff <= maxPopulation)
                    queue.addLast(new Coordinate(row + 1, col));
            }
            if (0 < col) {
                int populationDiff = Math.abs(map[row][col - 1] - map[row][col]);
                if (minPopulation <= populationDiff && populationDiff <= maxPopulation)
                    queue.addLast(new Coordinate(row, col - 1));
            }
            if (col < width - 1) {
                int populationDiff = Math.abs(map[row][col + 1] - map[row][col]);
                if (minPopulation <= populationDiff && populationDiff <= maxPopulation)
                    queue.addLast(new Coordinate(row, col + 1));
            }
        }

        if (union.size() > 1) {
            isOpen = true;
            int population = totalPopulation / numOfNations;
            for (Coordinate coordinate: union)
                map[coordinate.getRow()][coordinate.getCol()] = population;
        }
        return isOpen;
    }
}

class Coordinate {
    private final int row;
    private final int col;

    public Coordinate(int row, int col) { this.row = row; this.col = col; }

    public int getRow() { return row; }

    public int getCol() { return col; }
}
